pkg/bitlpm/cidr.go

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bitlpm
+
+import (
+	"math/bits"
+	"net/netip"
+)
+
+// CIDRTrie can hold both IPv4 and IPv6 prefixes
+// at the same time.
+type CIDRTrie[T any] struct {
+	v4 Trie[Key[netip.Prefix], T]
+	v6 Trie[Key[netip.Prefix], T]
+}
+
+// NewCIDRTrie creates a new CIDRTrie[T any].
+func NewCIDRTrie[T any]() *CIDRTrie[T] {
+	return &CIDRTrie[T]{
+		v4: NewTrie[netip.Prefix, T](32),
+		v6: NewTrie[netip.Prefix, T](128),
+	}
+}
+
+// Lookup returns the longest matched value for a given address.
+func (c *CIDRTrie[T]) Lookup(addr netip.Addr) T {
+	tri := c.v4
+	bits := 32
+	if addr.Is6() {
+		tri = c.v6
+		bits = 128
+	}
+	var def T
+	tri.Ancestors(uint(bits), cidrKey(netip.PrefixFrom(addr, bits)), func(_ uint, _ Key[netip.Prefix], v T) bool {
+		def = v
+		return true
+	})
+	return def
+}
+
+// Path calls fn for every entry that contains a given prefix.
+func (c *CIDRTrie[T]) Path(cidr netip.Prefix, fn func(k netip.Prefix, v T) bool) {
```

Comment:
As a user I'd find it useful to have an example to understand the bounds. For instance, is this accurate?

```suggestion
// Path calls fn for every entry that contains a given prefix.
//
// For example, t.Path(netip.MustParsePrefix("192.0.2.0/24")) would call 'fn' for each bitprefix from /0 up to and including the /24, corresponding to the following prefixes:
//  - "0.0.0.0/0"
//  - "128.0.0.0/1"
//  - "192.0.0.0/2"
//  - ...
//  - "192.0.2.0/24"
func (c *CIDRTrie[T]) Path(cidr netip.Prefix, fn func(k netip.Prefix, v T) bool) {
```

I also note that this does not mention the expected behaviour of `fn`. Perhaps it would make sense to create an explicit type for `fn` and in that type's comment, explain how the boolean value impacts control flow for iterator functions?

Reply:
Related to `fn` documentation: I assume it's bad if `fn` calls `trie.Delete(k)`?

Reply:
> Related to fn documentation: I assume it's bad if fn calls trie.Delete(k)?

Yes, that's what I'm trying to get at with the `non-thread-safe` term in the intro.

Reply:
Ah. I assumed non-thread-safe meant you can't edit it with two goroutines. I didn't think that applied to one goroutine making modifications during iteration.